File: #0273.integer-to-english-words.cpp
class Solution {
public:
    string numberToWords(int num) {

        vector<string> units = {
            "", " Thousand", " Million", " Billion"
        };

        int i = 0;
        string result;
        int remainder;
        while (num) {
            remainder = num % 1000;
            if (remainder) {
                result = numberWords(remainder) + units[i] + result;
            }
            i++;
            num /= 1000;
        }
        return result.length() ? result.substr(1) : "Zero";
    }

private:
    string numberWords(int num) {

        char* ones[] = {
            "", " One", " Two", " Three", " Four", " Five",
            " Six", " Seven", " Eight", " Nine"
        };
        char* tens[] = {
            " Ten", " Eleven", " Twelve", " Thirteen", " Fourteen",
            " Fifteen", " Sixteen", " Seventeen", " Eighteen", " Nineteen"
        };
        char* tenUnits[] = {
            "", "", " Twenty", " Thirty", " Forty", " Fifty",
            " Sixty", " Seventy", " Eighty", " Ninety"
        };

        string result;
        if (num > 99) {
            result += string(ones[num / 100]) + " Hundred";
        }

        num %= 100;
        if (num > 9 && num < 20) {
            result += tens[num % 10];
        } else {
            if (num > 19) {
                result += tenUnits[num / 10];
            }
            num %= 10;
            if (num) {
                result += ones[num];
            }
        }
        return result;
    }
};

File: #0273.integer-to-english-words.js
/**
 * @param {number} num
 * @return {string}
 */
var numberToWords = function(num) {
    if (num === 0) {
        return "Zero";
    }

    let str = "";
    const F = (x) => Math.floor(x);
    const dig = {
        1 : 'One',
        2 : 'Two',
        3 : 'Three',
        4 : 'Four',
        5 : 'Five',
        6 : 'Six',
        7 : 'Seven',
        8 : 'Eight',
        9 : 'Nine',
        10 : 'Ten',
        11 : 'Eleven',
        12 : 'Twelve',
        13 : 'Thirteen',
        14 : 'Fourteen',
        15 : 'Fifteen',
        16 : 'Sixteen',
        17 : 'Seventeen',
        18 : 'Eighteen',
        19 : 'Nineteen',
        20 : 'Twenty',
        30 : 'Thirty',
        40 : 'Forty',
        50 : 'Fifty',
        60 : 'Sixty',
        70 : 'Seventy',
        80 : 'Eighty',
        90 : 'Ninety'
    };

    const hundreds = (n) => {
        let temp = "";
        if (n >= 100) {
            temp += ` ${dig[F(n/100)]} Hundred`;
            n %= 100;
        }

        if (n >= 10 && n < 20) {
            temp += ` ${dig[n]}`;
            n = 0;
        }

        if (n >= 20) {
            temp += ` ${dig[F(n/10)*10]}`;
            n %= 10;
        }

        if (n) {
            temp += ` ${dig[F(n)]}`;
        }

        return temp;
    };

    if (num >= 1_000_000_000) {
        str += `${hundreds(F(num/1_000_000_000))} Billion`;
        num %= 1_000_000_000;
    }

    if (num >= 1_000_000) {
        str += `${hundreds(F(num/1_000_000))} Million`;
        num %= 1_000_000;
    }

    if (num >= 1_000) {
        str += `${hundreds(F(num/1_000))} Thousand`;
        num %= 1_000;
    }

    if (num) {
        str += `${hundreds(F(num))}`;
    }
    
    return str.trim();
};
